[MIG][11.0] web_widget_datepicker_options: Migrate module to v11

yaniaular opened the pull request:
This migrates the module [web_widget_datepicker_options](https://github.com/OCA/web/tree/10.0/web_widget_datepicker_options) which had not been migrated yet from v10.

This applies some changes to the module, to conform new guidelines and
conventions for version 11.0.

In addition, this applies the following fixes:
- The widged didn't work when the field was a `datetime` due to a typo
  in a JS conditional
- Replaced `Object.assign()` -> `_.assign()` to improve
  browser-compatibility

Commit:
[ADD] web_widget_datepicker_options

* use include instead of extend

* add README.rst

* update api doc link in readme.rst

* remove description from __oe__ since it is in README.rst

* add OCA icon and AGPL badges

* add datepicker.js through assets_backend

yvaucher committed:
Add bug tracker link on README.rst

Commit:
add OCA to author

sbidoul committed:
[UPD] prefix versions with 8.0

pedrobaeza committed:
[MIG] Make modules uninstallable

pedrobaeza committed:
[MIG] Rename manifest files

Commit:
[MIG] Migrate web_widget_datepicker_options to version 11.0

This commit extracts the module from version 10 and migrates it to v11

yaniaular commented:
@veloutin @yvaucher @sbidoul @pedrobaeza I have migrated this module and I need your approval to merge it, thanks.

pedrobaeza commented:
Hi @yaniaular, thanks for the migration, but I'm afraid this is not how it works. Reviewers can come to this PR because they are interested in the module itself, or you can review other PRs in the same repository or other similar ones and ask in exchange that they review yours. Current commit authors that you see can be general OCA maintainers like me and @sbidoul that don't know about the module itself, or even non active Odoo developers.

yaniaular commented:
@pedrobaeza I think you get me wrong. I didn't mean I was going to merge this. I just wanted to tell you that I've migrated the module and that you may review it. I do not have permission to merge it anyway.

simahawk commented on web_widget_datepicker_options/README.rst:
```diff
@@ -0,0 +1,59 @@
+.. image:: https://img.shields.io/badge/licence-AGPL--3-blue.svg
```
png

simahawk commented on web_widget_datepicker_options/README.rst:
```diff
@@ -0,0 +1,59 @@
+* Absolutely no validation on options.
+
+
+Bug Tracker
```
bug tracker and credits need update from template https://github.com/OCA/maintainer-tools/blob/master/template/module/README.rst

simahawk commented on web_widget_datepicker_options/__init__.py:
```diff
@@ -0,0 +1,21 @@
+# -*- encoding: utf-8 -*-
```
remove encoding and use short license headers in all py files

simahawk commented on web_widget_datepicker_options/__manifest__.py:
```diff
@@ -0,0 +1,37 @@
+{
+    "name": 'web_widget_datepicker_options',
+    "version": "11.0.1.0.0",
+    "license": "AGPL-3",
```
`'website': 'https://github.com/OCA/web',`

simahawk commented on web_widget_datepicker_options/__manifest__.py:
```diff
@@ -0,0 +1,37 @@
+##############################################################################
+
+{
+    "name": 'web_widget_datepicker_options',
```
```
'name': 'Web widget datepicker options',
'summary': 'Enhance customization for datepicker widgets',
```

simahawk commented on web_widget_datepicker_options/static/src/js/datepicker.js:
```diff
@@ -0,0 +1,43 @@
+    Widget.DateWidget.include({
+        init: function() {
+            this._super.apply(this, arguments);
+            if(typeof this.__parentedParent !== 'undefined' && this.__parentedParent.field.type === 'date' && this.__parentedParent.nodeOptions){
```
define `var parent = this.getParent();` and replace all `this.__parentedParent`

simahawk commented on web_widget_datepicker_options/view/qweb.xml:
```diff
@@ -0,0 +1,10 @@
+<openerp>
```
s/openerp/odoo
remove `<data />`

simahawk commented on web_widget_datepicker_options/README.rst:
```diff
@@ -0,0 +1,59 @@
+You must pass all options through the "datepicker" field in the options::
+
+    ...
+    <field name="date" options="{'datepicker':{'yearRange': 'c-100:c+0'}}"/>
```
Am I wrong or you should use JSON notation?
```xml
<field name="date" options='{"datepicker":{"yearRange": "c-100:c+0"}}'/>
```

luisg123v replied:
@simahawk 
Both single and double quotes are possible, it keeps the same behavior as Odoo's native options.

MeetKD commented on web_widget_datepicker_options/__manifest__.py:
```diff
@@ -0,0 +1,37 @@
+    "depends": [
+        'base',
+        'web',
+    ],
```
Add only web module dependency.

luisg123v committed:
[FIX] web_widget_datepicker_options: Apply fixes after migration to v11

This applies some changes to the module after migration, to conform new
guidelines and conventions for version 11.0.

In addition, this applies the following fixes:
- The widged didn't work when the field was a `datetime` due to a typo
  in a JS conditional
- Replaced `Object.assign()` -> `_.assign()` to improve
  browser-compatibility

luisg123v commented:
@simahawk & @MeetKD 
We applied all your requested changes. Also we edited the PR's description, to make clearer what this is all about.

If it's not too much, could you review again, please?

simahawk reviewed (approved):
LGTM. You could squash last 2 commits before merge.

luisg123v commented:
@simahawk 
Are you sure squashing is the way to go? Take into account the author differs for those commits.

luisg123v commented:
@moylop260 
Could you approve this PR, please?

moylop260 commented:
Could you squash the Yanina's commits in one commit, please?


UPDATED:
I mean, the following commits
 - https://github.com/OCA/web/pull/924/commits/68d4bde228e6983c1cf9f6af9a09e89901623899
 - https://github.com/OCA/web/pull/924/commits/c00ebff968286d7bddd4ab4288a6ce603b870c21

luisg123v commented:
@moylop260 
Done, commits were squashed as you requested

yajo reviewed (commented):
I might be saying some stupidity, but [isn't this already what the v11 core widget does](https://github.com/odoo/odoo/blob/1039abfbdb1dcfd51e753874795721fc0d617485/addons/web/static/src/js/widgets/date_picker.js#L22-L57)?

luisg123v commented:
@Yajo 
AFAIK, it's not the same. That method merge a parameter named `options`, with some computed values. However, that parameter is not the options passed to the field in a view, but some common values that Odoo passes to all widgets

If you wan to, e.g. limit a field, the following won't work:
```
<field name="post_date" options="{'daysOfWeekDisabled': [0, 6]}"/>
```
But, with this module, you could do:
```
<field name="post_date" options="{'datepicker': {'daysOfWeekDisabled': [0, 6]}}"/>
```

pedrobaeza commented:
I think first 8 commits from same author can be squashed, as that commit history doesn't add anything useful. When done, I'll merge.

yaniaular commented:
@pedrobaeza Do you mean the veloutin's commits?

pedrobaeza commented:
Yeah, that's it

luisg123v commented:
@pedrobaeza 
Done, all commits from the same author were squashed.

yajo commented:
Fixing this upstream is 1 LOC, I did it here: https://github.com/odoo/odoo/pull/25044

Let's hope they merge it and we need no more this addon :blush:
